### PR TITLE
[manila]: Add Share Replicas support

### LINKED ISSUE #2633: [manila]: implement share replicas API
https://docs.openstack.org/api-ref/shared-file-system/#share-replicas-since-api-v2-11

Share replicas are the replicated copies of the existing share. You can use Share Replicas to sync data so that each share replica has an identical copy of the same share. Share replication can be used as a disaster recovery solution or as a load sharing mirroring solution.

Manila supports replication of shares between different storage pools. These pools may be on different back-end storage systems or within the same back end, depending upon the replication style chosen, the capability of the driver and the configuration of back ends. To ensure that a secondary copy is scheduled to a distinct back end, you must specify the availability_zone attribute.

More details: https://docs.openstack.org/manila/latest/contributor/share_replication.html

### PR DESCRIPTION
**Pull request by kayrus**
Fixes #2633 

Looks like devstack doesn't support replicas. Here is the output of local tests:

```
$ go test -run TestReplica -v ./*.go
=== RUN   TestReplicaCreate
    tools.go:80: {
          "id": "125d64dd-aab9-4e24-8ccb-39dd455d1ba7",
          "availability_zone": "zone-1b",
          "cast_rules_to_readonly": false,
          "host": "",
          "share_id": "72a37fa6-20c9-43dc-a463-b802e96044d0",
          "share_network_id": "ca0163c8-3941-4420-8b01-41517e19e366",
          "share_server_id": "ed15ae66-af5d-4d72-8db2-a1f90a93a260",
          "status": "available",
          "replica_state": "out_of_sync"
        }
    replicas.go:54: Deleted replica: 125d64dd-aab9-4e24-8ccb-39dd455d1ba7
    shares.go:103: Deleted share: 72a37fa6-20c9-43dc-a463-b802e96044d0
--- PASS: TestReplicaCreate (66.88s)
=== RUN   TestReplicaPromote
    tools.go:80: {
          "id": "6a39a458-b2ee-468f-b1d9-261db50eec8c",
          "availability_zone": "zone-1b",
          "cast_rules_to_readonly": false,
          "host": "",
          "share_id": "fe52148b-3a24-4a19-84aa-08f1b2eeafd8",
          "share_network_id": "ca0163c8-3941-4420-8b01-41517e19e366",
          "share_server_id": "ed15ae66-af5d-4d72-8db2-a1f90a93a260",
          "status": "available",
          "replica_state": "out_of_sync"
        }
    replicas.go:54: Deleted replica: 6a39a458-b2ee-468f-b1d9-261db50eec8c
    shares.go:103: Deleted share: fe52148b-3a24-4a19-84aa-08f1b2eeafd8
--- PASS: TestReplicaPromote (126.09s)
=== RUN   TestReplicaExportLocations
    tools.go:80: []
    tools.go:80: [
          {
            "id": "c0cee115-28fc-43f0-bb42-5eba7f09d5dc",
            "path": "10.180.66.129:/share_5b891aba_b9a9_4582_8fb3_e14ab0dc2b6c",
            "share_instance_id": "",
            "is_admin_only": false,
            "preferred": true,
            "availability_zone": "zone-1b",
            "replica_state": "active"
          },
          {
            "id": "97bcdc70-ca37-45e9-b02b-3918b035c69b",
            "path": "10.180.66.22:/share_5b891aba_b9a9_4582_8fb3_e14ab0dc2b6c",
            "share_instance_id": "",
            "is_admin_only": false,
            "preferred": false,
            "availability_zone": "zone-1b",
            "replica_state": "active"
          }
        ]
    tools.go:80: {
          "id": "c0cee115-28fc-43f0-bb42-5eba7f09d5dc",
          "path": "10.180.66.129:/share_5b891aba_b9a9_4582_8fb3_e14ab0dc2b6c",
          "share_instance_id": "",
          "is_admin_only": false,
          "preferred": true,
          "availability_zone": "zone-1b",
          "replica_state": "active"
        }
    replicas.go:54: Deleted replica: 53661add-3fa1-49a2-b482-45c8c16a3e0a
    shares.go:103: Deleted share: 66c839d2-b083-4cb7-a4f8-eab71e43c908
--- PASS: TestReplicaExportLocations (54.36s)
=== RUN   TestReplicaListDetail
    tools.go:80: {
          "id": "65fe1244-ac6f-4e73-80e4-597a95ec6520",
          "availability_zone": "",
          "cast_rules_to_readonly": false,
          "host": "",
          "share_id": "e836fc36-62ef-4c82-b873-7c9fac061819",
          "share_network_id": "",
          "share_server_id": "",
          "status": "available",
          "replica_state": "out_of_sync"
        }
    tools.go:80: {
          "id": "f2e24517-87e5-4b35-b594-f69c2bb09ab4",
          "availability_zone": "",
          "cast_rules_to_readonly": false,
          "host": "",
          "share_id": "e836fc36-62ef-4c82-b873-7c9fac061819",
          "share_network_id": "",
          "share_server_id": "",
          "status": "available",
          "replica_state": "active"
        }
    replicas.go:54: Deleted replica: 65fe1244-ac6f-4e73-80e4-597a95ec6520
    shares.go:103: Deleted share: e836fc36-62ef-4c82-b873-7c9fac061819
--- PASS: TestReplicaListDetail (66.38s)
=== RUN   TestReplicaResetStatus
    replicas_test.go:278: Replica bf6d605c-3773-4178-8ae1-3b177540ffd0 status successfuly reset
    replicas.go:54: Deleted replica: bf6d605c-3773-4178-8ae1-3b177540ffd0
    shares.go:103: Deleted share: ef9fe330-4a4f-4703-ba97-a853d70a8314
--- PASS: TestReplicaResetStatus (80.83s)
=== RUN   TestReplicaForceDelete
    replicas_test.go:23: Skipping cloud admin tests.
--- SKIP: TestReplicaForceDelete (0.00s)
PASS
ok      command-line-arguments  394.539s
```

### CODE REVIEW

**Comment by kayrus** [comment]
@mandre @pierreprinetti ready for review

**Comment by kayrus** [comment]
@mandre @pierreprinetti kindly ping

**Comment by kayrus** [comment]
any updates on this PR?

**Comment by mandre** [comment]
> any updates on this PR?

About to take a look. Stay tuned :)

File: acceptance/openstack/sharedfilesystems/v2/replicas_test.go [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+	"github.com/gophercloud/gophercloud/acceptance/clients"
+	"github.com/gophercloud/gophercloud/acceptance/tools"
+	"github.com/gophercloud/gophercloud/openstack/sharedfilesystems/v2/replicas"
+	th "github.com/gophercloud/gophercloud/testhelper"
+)
+
+// otherwise we need to set "X-OpenStack-Manila-API-Experimental: true"
+const replicasMicroversion = "2.60"
+
+func skipReplicaTests(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mandre) [reply acceptance/openstack/sharedfilesystems/v2/replicas_test.go:19]: I suspect adding `replication_type=readable` to `MANILA_DEFAULT_SHARE_TYPE_EXTRA_SPECS` might do the trick.

https://github.com/gophercloud/gophercloud/blob/master/.github/workflows/functional-sharedfilesystems.yaml#LL54C13-L54C50

File: openstack/sharedfilesystems/v2/replicas/results.go [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+)
+
+// Replica contains all information associated with an OpenStack Replica Replica
+type Replica struct {
+	// ID of the share replica
+	ID string `json:"id"`
+	// The availability zone of the share replica
+	AvailabilityZone string `json:"availability_zone"`
+	// Indicates whether existing access rules will be cast to read/only
+	CastRulesToReadonly bool `json:"cast_rules_to_readonly"`
```
[the remark this answers is not in the document]
Comment (kayrus) [reply openstack/sharedfilesystems/v2/replicas/results.go:24]: IMHO, adding an extra struct type would add even more confusion, especially when the results may be used within a single scope. BTW, the same bool issue can be faced with the `is_admin_only` field in the `ExportLocation` struct:

> This parameter is only available to users with an “administrator” role, and cannot be controlled via policy .json.

Maybe it makes sense to use pointers to indicate whether the bool value is set?

Also, listing replica export locations doesn't return created/updated timestamps, while getting a single export location does. I don't think that we need to define an extra struct for this case as well.

File: acceptance/openstack/sharedfilesystems/v2/replicas_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kayrus) [reply acceptance/openstack/sharedfilesystems/v2/replicas_test.go:19]: Unfortunately I haven't managed to enable the replication (which requires ZFS on Linux) support. I disabled it again.

**Comment by kayrus** [comment]
@mandre is there anything else I can do?

File: acceptance/openstack/sharedfilesystems/v2/replicas_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kayrus) [reply acceptance/openstack/sharedfilesystems/v2/replicas_test.go]: I added an `OS_MANILA_REPLICAS` env check to run manila replicas acceptance tests.

**Comment by mandre** [comment]
> @mandre is there anything else I can do?

No, it's waiting on me, sorry. Just haven't had the chance to get back to reviewing your PR yet.

In your latest change, what sets OS_MANILA_REPLICAS?

**Review by mandre — APPROVED** [review]

**Comment by kayrus** [comment]
@mandre we can try to setup the env with zfs that supports replicas later, but IMO it won't be easy. Locally I test with NetApp backend, it works great.

**Pull request merged**